### ISSUE TITLE
`tsc --build` errors on first run, passes on second run

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#44605]
# Bug Report

### 🔎 Search Terms

project references build

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `tsc --build`

### ⏯ Playground Link

N/A

### 💻 Code

1. Clone and checkout https://github.com/unsplash/unsplash-web/tree/typescript-tsc-build-cache-issue, branch `typescript-tsc-build-cache-issue`. This is a private repository but I believe some members of the TS team already have access from debugging previous issues. If you don't have access, please send me your email address so I can provide access (TS team members only).
2. Run `yarn` to install TS
3. Run these commands:
   ```bash
   # This simulates a cache inside of our CI builds
   rm -rf target-tsc-build && cp -r target-tsc-build-copy target-tsc-build
   # don't ask me why
   touch shared/facades/io-ts/operators.ts

   tsc --build tsconfig.app.json --verbose
   ```
4. After verifying there were 2 type errors, run `tsc --build tsconfig.app.json --verbose` again. 

### 🙁 Actual behavior

`tsc --build` errors on first run, passes on second run

### 🙂 Expected behavior

Consistent behaviour:

- `tsc --build` errors on first run and second run
- OR, `tsc --build` passes on first run and second run

**Comment by RyanCavanaugh** [comment]
Can I have access 🙂

**Comment by OliverJAsh** [comment]
Done!

**Comment by RyanCavanaugh** [comment]
This isn't reproing for me - might be a Windows vs non-Windows thing. Can you email me the console printback from your repro?

**Comment by RyanCavanaugh** [comment]
Or maybe better, point me at an incorrectly-passing Travis run

**Comment by OliverJAsh** [comment]
Unfortunately I can't link to the Travis build (which was incorrectly failing rather than incorrectly passing), but I can post the error here.

```
$ npm start --silent checkCi.$TRAVIS_STEP
nps is executing `checkCi.typecheck` : nps typecheck.shared && nps typecheck.app && nps typecheck.webpack && nps typecheck.cypress && nps typecheck.service-worker && nps typecheck.build-tracker
nps is executing `typecheck.shared` : tsc --build ./shared/tsconfig.json --verbose
8:30:50 PM - Projects in this build: 
    * shared/tsconfig.json
8:30:50 PM - Project 'shared/tsconfig.json' is out of date because oldest output 'target-tsc-build/shared/RailsOAuth.js' is older than newest input 'shared/typings-base/node-process-env-augmentations.d.ts'
8:30:50 PM - Building project '/home/travis/build/unsplash/unsplash-web/shared/tsconfig.json'...
8:31:00 PM - Updating unchanged output timestamps of project '/home/travis/build/unsplash/unsplash-web/shared/tsconfig.json'...
nps is executing `typecheck.app` : tsc --build ./tsconfig.app.json --verbose
8:31:00 PM - Projects in this build: 
    * shared/tsconfig.json
    * tsconfig.app.json
8:31:00 PM - Project 'shared/tsconfig.json' is up to date because newest input 'shared/typings-base/node-process-env-augmentations.d.ts' is older than oldest output 'target-tsc-build/shared/tracking.d.ts'
8:31:00 PM - Project 'tsconfig.app.json' is out of date because oldest output 'target-tsc-build/setup-env-vars.js' is older than newest input 'shared/tsconfig.json'
8:31:00 PM - Building project '/home/travis/build/unsplash/unsplash-web/tsconfig.app.json'...
app/routes/CampaignProposal/data-fetching.ts(69,21): error TS2345: Argument of type '([response, originalResponse]: WithResponse<CampaignProposalProfileResponse>) => Reader<RequestFunction, Observable<Left<({ tag: "FetchError"; } & { ...; }) | ({ ...; } & APIError) | ({ ...; } & { ...; }) | ({ ...; } & { ...; })> | Right<...> | Right<...>>>' is not assignable to parameter of type '(a: WithResponse<CampaignProposalProfileResponse>) => ReaderObservableEither<RequestFunction, ({ tag: "FetchError"; } & { error: FetchError; }) | ({ ...; } & APIError) | ({ ...; } & { ...; }) | ({ ...; } & { ...; }), [...]>'.
  Call signature return types 'Reader<RequestFunction, Observable<Left<({ tag: "FetchError"; } & { error: FetchError; }) | ({ tag: "APIError"; } & APIError) | ({ tag: "UnknownAPIError"; } & { ...; }) | ({ ...; } & { ...; })> | Right<...> | Right<...>>>' and 'ReaderObservableEither<RequestFunction, ({ tag: "FetchError"; } & { error: FetchError; }) | ({ tag: "APIError"; } & APIError) | ({ tag: "UnknownAPIError"; } & { message: string; }) | ({ ...; } & { ...; }), [...]>' are incompatible.
    Call signature return types 'Observable<Left<({ tag: "FetchError"; } & { error: FetchError; }) | ({ tag: "APIError"; } & APIError) | ({ tag: "UnknownAPIError"; } & { message: string; }) | ({ tag: "HandleResponseError"; } & { ...; })> | Right<...> | Right<...>>' and 'ObservableEither<({ tag: "FetchError"; } & { error: FetchError; }) | ({ tag: "APIError"; } & APIError) | ({ tag: "UnknownAPIError"; } & { message: string; }) | ({ tag: "HandleResponseError"; } & { ...; }), [...]>' are incompatible.
      The types of 'subscribe' are incompatible between these types.
        Type '{ (observer?: PartialObserver<Left<({ tag: "FetchError"; } & { error: FetchError; }) | ({ tag: "APIError"; } & APIError) | ({ tag: "UnknownAPIError"; } & { message: string; }) | ({ tag: "HandleResponseError"; } & { ...; })> | Right<...> | Right<...>> | undefined): Subscription; (next: null | undefined, error: null |...' is not assignable to type '{ (observer?: PartialObserver<Either<({ tag: "FetchError"; } & { error: FetchError; }) | ({ tag: "APIError"; } & APIError) | ({ tag: "UnknownAPIError"; } & { message: string; }) | ({ tag: "HandleResponseError"; } & { ...; }), [...]>> | undefined): Subscription; (next: null | undefined, error: null | undefined, compl...'.
          Types of parameters 'next' and 'next' are incompatible.
            Types of parameters 'value' and 'value' are incompatible.
              Type 'Left<({ tag: "FetchError"; } & { error: FetchError; }) | ({ tag: "APIError"; } & APIError) | ({ tag: "UnknownAPIError"; } & { message: string; }) | ({ tag: "HandleResponseError"; } & { ...; })> | Right<...> | Right<...>' is not assignable to type 'Either<({ tag: "FetchError"; } & { error: FetchError; }) | ({ tag: "APIError"; } & APIError) | ({ tag: "UnknownAPIError"; } & { message: string; }) | ({ tag: "HandleResponseError"; } & { ...; }), [...]>'.
                Type 'Right<ActionsWithResponse>' is not assignable to type 'Either<({ tag: "FetchError"; } & { error: FetchError; }) | ({ tag: "APIError"; } & APIError) | ({ tag: "UnknownAPIError"; } & { message: string; }) | ({ tag: "HandleResponseError"; } & { ...; }), [...]>'.
                  Type 'Right<ActionsWithResponse>' is not assignable to type 'Right<[Action<any>[], Response]>'.
                    Type 'ActionsWithResponse' is not assignable to type '[Action<any>[], Response]'.
                      Type 'Response | undefined' is not assignable to type 'Response'.
                        Type 'undefined' is not assignable to type 'Response'.
app/helpers/data-fetching.ts(115,7): error TS2322: Type '({ companyId, }: Route.Data.UnionRecord['CampaignProposal']) => ApiReader.ReaderObservableEither<ApiReader.RequestFunction, ({ tag: "FetchError"; } & { error: FetchError; }) | ({ tag: "APIError"; } & APIError) | ({ tag: "UnknownAPIError"; } & { ...; }) | ({ ...; } & { ...; }), unknown>' is not assignable to type '(value: { companyId: string; companyName: string; }) => BuildRouteActions'.
  Type 'ReaderObservableEither<RequestFunction, ({ tag: "FetchError"; } & { error: FetchError; }) | ({ tag: "APIError"; } & APIError) | ({ tag: "UnknownAPIError"; } & { message: string; }) | ({ ...; } & { ...; }), unknown>' is not assignable to type 'BuildRouteActions'.
    Type 'unknown' is not assignable to type 'ActionsWithResponse'.
```

**Comment by RyanCavanaugh** [comment]
I've narrowed this down to a file ordering dependence. What's happening is that the OS is returning a different order of files during directory traversal on the subsequent builds, so tsc processes them in a different order, and then we have some bad corner case somewhere that causes typechecking to behave differently depending on the results.

Rough process of investigation here:
 * Run `tsc -p tsconfig.app.json --noEmit --listFiles`
 * Dump the emitted file listing into the `files` array of `tsconfig.app.json`
 * Remove `apply-test-env-vars.d.ts` from the list because this is somehow in the program by another name (??)
 * Run `tsc -p tsconfig.app.json --noEmit` again, you should get the same result as in step 1
 * Reverse the ordering of the files list starting at the first non-lib file
 * Run `tsc -p tsconfig.app.json --noEmit` again, the error state should flip (2 becomes 0, 0 becomes 2)

*In theory*, there's a single file somewhere in your build that you can put in the `files` array to fix the ordering to the desired one. Figuring out *which* one is going to take some codebase-specific knowledge and digging.

Once you've determined that, it should become tractable to reduce this to a repro. File ordering typechecking dependencies are always a bug, so at that point we can fix the issue.

**Comment by RyanCavanaugh** [comment]
Two other notes:
* There's a "Flip" extension for VS Code for reordering lines 🙂
* This is in-principle findable by doing a binary search on which half of the files list to reverse, *I think*?

**Comment by OliverJAsh** [comment]
> * Run `tsc -p tsconfig.app.json --noEmit` again, the error state should flip (2 becomes 0, 0 becomes 2)

When I followed these steps, the error state didn't flip—I get 2 errors before I reverse the ordering of the files list starting at the first non-lib file, and I still get 2 errors after the reverse.

> * Remove `apply-test-env-vars.d.ts` from the list because this is somehow in the program by another name (??)

In the printed list of files I see `/Users/oliverash/Development/unsplash-web-temp/target-tsc-build/shared/apply-test-env-vars.d.ts` which is correct. What was the other name you were seeing? (FWIW I removed it from the file list anyway.)
